### ISSUE TITLE
[BUG] The same data in multiple *-storage.json 

### ISSUE THREAD

**Issue by apotrebin** [microsoft/playwright#24540]
### System info
- Playwright Version: [v1.36.2]
- Operating System: [macOS 13.5, ubuntu 20]
- Browser: [All]

**Config file**

```js
// playwright.config.ts
projects: [
    { 
      name: 'setup',
      testMatch: /.*\.setup\.ts/ 
    },

    {
      name: 'chromium',
      use: {
        ...devices['Desktop Chrome'],
        browserName: 'chromium',
        // storageState: 'storage/admin-api.json'
      },
      dependencies: ['setup']
    },
```

**Test file (self-contained)**

```js
// auth.setup.ts
import { test as setup } from '@playwright/test'

setup('auth setup', async ({ request } ) => {
    switch (ENV) {
      case 'dev':
          await authAs({request}, 'dev', 'admin', process.env.ADMIN_ROLE, process.env.ACCOUNT_PASSWORD)
          await authAs({request}, 'dev', 'user', process.env.USER_ROLE, process.env.ACCOUNT_PASSWORD)
          await authAs({request}, 'dev', 'it_support', process.env.IT_SUPPORT_ROLE, process.env.ACCOUNT_PASSWORD)
        break
      default:
        console.log("WARNING! ENV is not defined")
        break
    }
})

async function authAs({request}, env: String, role: String, account: String, password: String) {
  const response = await request.post(`https://api-${env}-***.com/auth/login`, { form: { 'username': `${account}`, 'password': `${password}` } })
  await request.storageState({path: `storage/${role}-api.json`})
  console.log(`=> auth status: [${response.status()}] on [${env}] for [ ${role} ] `) 
}
```

```js
// any test file 
test.use({ storageState: './storage/admin-api.json' })
```

**Steps**
- [Run the test]
- [I get 3 files: storage/admin-api.json , storage/user-api.json, storage/it_support-api.json]

**Expected**
All 3 files (storage/admin-api.json , storage/user-api.json, storage/it_support-api.json) contains different cookie's value

**Actual**
All 3 files contains the same data and data is equal of the latest run.
 In my case latest run is 
```await authAs({request}, 'dev', 'it_support', process.env.IT_SUPPORT_ROLE, process.env.ACCOUNT_PASSWORD)```
 In other words storage/admin-api.json will contain data of storage/it_support-api.json

I am not sure if it a bug of PW , cause I'm not good at typescript, but similar code worked well when I used auth using UI

P.S. The main idea of mine was to create several storage/${role}-api.json before all kind of tests and use particular role further in test I need.

**Comment by yury-s** [comment]
You solution above is good. The problem is you were reusing the same APIRequestContext for all 3 users. Alternatively you could have 3 `setup`, one for each role and make them have separate `request` instances:

```ts

setup('auth admin', async ({ playwright } ) => {
    switch (ENV) {
      case 'dev':
          await authAs({playwright}, 'dev', 'admin', process.env.ADMIN_ROLE, process.env.ACCOUNT_PASSWORD)
        break
      default:
        console.log("WARNING! ENV is not defined")
        break
    }
})

setup('auth user', async ({ playwright } ) => {
    switch (ENV) {
      case 'dev':
          await authAs({playwright}, 'dev', 'user', process.env.USER_ROLE, process.env.ACCOUNT_PASSWORD)
        break
      default:
        console.log("WARNING! ENV is not defined")
        break
    }
})

setup('auth it_support', async ({ playwright } ) => {
    switch (ENV) {
      case 'dev':
          await authAs({playwright}, 'dev', 'it_support', process.env.IT_SUPPORT_ROLE, process.env.ACCOUNT_PASSWORD)
        break
      default:
        console.log("WARNING! ENV is not defined")
        break
    }
})
```
